tinygraph/tinygraph-index.c

```diff
@@ -0,0 +1,424 @@
+      .lng = lngs[i],
+      .lat = lats[i],
+    };
+
+    items[i] = item;
+  }
+
+  qsort(items, n, sizeof(tinygraph_index_item), tinygraph_index_item_cmp);
+
+  for (uint32_t i = 0; i < n; ++i) {
+    nodes_[i] = items[i].node;
+    zvals_[i] = items[i].zval;
+    lngs_[i] = items[i].lng;
+    lats_[i] = items[i].lat;
+  }
+
+  free(items);
+
+  *out = (tinygraph_index) {
+    .size = n,
+    .nodes = nodes_,
+    .zvals = zvals_,
+    .lngs = lngs_,
+    .lats = lats_,
+  };
+
+  return out;
+}
+
+
+tinygraph_index* tinygraph_index_copy(const tinygraph_index * const index) {
+  TINYGRAPH_ASSERT(index);
+  TINYGRAPH_ASSERT(index->nodes);
+  TINYGRAPH_ASSERT(index->zvals);
+  TINYGRAPH_ASSERT(index->lngs);
+  TINYGRAPH_ASSERT(index->lats);
+
+  tinygraph_index *copy = malloc(sizeof(tinygraph_index));
+
+  if (!copy) {
+    return NULL;
+  }
+
+  uint32_t* nodes = malloc(index->size * sizeof(uint32_t));
+
+  if (!nodes) {
+    free(copy);
+
+    return NULL;
+  }
+
+  uint64_t* zvals = malloc(index->size * sizeof(uint64_t));
+
+  if (!zvals) {
+    free(nodes);
+    free(copy);
+
+    return NULL;
+  }
+
+  uint32_t* lngs = malloc(index->size * sizeof(uint32_t));
+
+  if (!lngs) {
+    free(zvals);
+    free(nodes);
+    free(copy);
+
+    return NULL;
+  }
+
+  uint32_t* lats = malloc(index->size * sizeof(uint32_t));
+
+  if (!lats) {
+    free(lngs);
+    free(zvals);
+    free(nodes);
+    free(copy);
+
+    return NULL;
+  }
+
+  memcpy(nodes, index->nodes, index->size * sizeof(uint32_t));
+  memcpy(zvals, index->zvals, index->size * sizeof(uint64_t));
+  memcpy(lngs, index->lngs, index->size * sizeof(uint32_t));
+  memcpy(lats, index->lats, index->size * sizeof(uint32_t));
+
+  *copy = (tinygraph_index) {
+    .size = index->size,
+    .nodes = nodes,
+    .zvals = zvals,
+    .lngs = lngs,
+    .lats = lats,
+  };
+
+  return copy;
+}
+
+
+void tinygraph_index_destruct(tinygraph_index * const index) {
+  if (!index) {
+    return;
+  }
+
+  free(index->lats);
+  free(index->lngs);
+  free(index->zvals);
+  free(index->nodes);
+
+  free(index);
+}
+
+
+static inline const uint64_t* tinygraph_index_bsearch_lt(
+    const uint64_t* first,
+    const uint64_t* last,
+    uint64_t value
+) {
+  const uint64_t *it = first;
+
+  uint64_t step = 0, n = last - first;
+
+  while (n > 0) {
+    step = n / 2;
+    it = first + step;
+
+    if (*it < value) { // <- lt means less than here
+      first = it + 1;
+      n -= step + 1;
+    } else {
+      n = step;
+    }
+  }
+
+  return first;
+}
+
+static inline const uint64_t* tinygraph_index_bsearch_lte(
+    const uint64_t* first,
+    const uint64_t* last,
+    uint64_t value
+) {
+  const uint64_t *it = first;
+
+  uint64_t step = 0, n = last - first;
+
+  while (n > 0) {
+    step = n / 2;
+    it = first + step;
+
+    if (*it <= value) { // <- lte means less than or equal here
+      first = it + 1;
+      n -= step + 1;
+    } else {
+      n = step;
+    }
+  }
+
+  return first;
+}
```

Comment:
Simple binary search implementations to get something started. We need a `<` and a `<=` implement for the range searches along the space filling z-order curve below.